playground/custom-server-single-port/__tests__/custom-server-single-port.spec.ts

```diff
@@ -0,0 +1,34 @@
+import { expect, test } from 'vitest'
+import { port } from './serve'
+import { page } from '~utils'
+
+const url = `http://localhost:${port}/`
+
+const waitForInitialHmrWebSocketMessageWithoutRequestFailure = () =>
+  new Promise<{ webSocketUrl: string; payload: any }>((resolve, reject) => {
+    page.on('requestfailed', (request) => {
+      reject(new Error(request.failure().errorText + ` (${request.url()})`))
+    })
+
+    page.once('websocket', (ws) => {
+      ws.once('framereceived', (data) => {
+        resolve({
+          webSocketUrl: ws.url(),
+          payload: JSON.parse(data.payload.toString()),
+        })
+      })
+      ws.once('socketerror', (err) => {
+        reject(err)
+      })
+    })
+  })
+
+test('HMR WebSocket connection should use the same host as the main server', async () => {
```

Comment:
Note: I confirmed that this added test case fails as expected on the current `main` branch (13ac37df8e135fe6b4990fac6f149a36a19e5409) and it passes with this PR.
Here is the local test run log:

### On `main` (13ac37df8e135fe6b4990fac6f149a36a19e5409)

<details><summary>Test run output</summary>
    $ pnpm run test-serve custom-server-single-port.spec.ts
    
    > @vitejs/vite-monorepo@ test-serve /Users/naru/github.com/vitejs/vite
    > vitest run -c vitest.config.e2e.ts "custom-server-single-port.spec.ts"
    
    
     RUN  v0.25.7 /Users/naru/github.com/vitejs/vite
    
    x
    
    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    
     FAIL  playground/custom-server-single-port/__tests__/custom-server-single-port.spec.ts > HMR WebSocket connection should use the same host as the main server
    Error: net::ERR_CONNECTION_REFUSED (http://localhost:24678/)
     ❯ Page.<anonymous> playground/custom-server-single-port/__tests__/custom-server-single-port.spec.ts:10:14
          8|   new Promise<{ webSocketUrl: string; payload: any }>((resolve, reject) => {
          9|     page.on('requestfailed', (request) => {
         10|       reject(new Error(request.failure().errorText + ` (${request.url()})`))
           |              ^
         11|     })
         12|
     ❯ Page.emit node:events:513:28
     ❯ BrowserContext._onRequestFailed node_modules/.pnpm/playwright-core@1.28.1/node_modules/playwright-core/lib/client/browserContext.js:146:20
     ❯ Proxy.<anonymous> node_modules/.pnpm/playwright-core@1.28.1/node_modules/playwright-core/lib/client/browserContext.js:112:16
     ❯ Proxy.emit node:events:513:28
     ❯ Connection.dispatch node_modules/.pnpm/playwright-core@1.28.1/node_modules/playwright-core/lib/client/connection.js:178:21
     ❯ Proxy.<anonymous> node_modules/.pnpm/playwright-core@1.28.1/node_modules/playwright-core/lib/client/browserType.js:176:22
     ❯ Proxy.emit node:events:513:28
     ❯ Connection.dispatch node_modules/.pnpm/playwright-core@1.28.1/node_modules/playwright-core/lib/client/connection.js:178:21
     ❯ Immediate.<anonymous> node_modules/.pnpm/playwright-core@1.28.1/node_modules/playwright-core/lib/inProcessFactory.js:46:83
</details>


### This PR

<details><summary>Test run output</summary>
    $ pnpm run test-serve custom-server-single-port.spec.ts
    
    > @vitejs/vite-monorepo@ test-serve /Users/naru/github.com/vitejs/vite
    > vitest run -c vitest.config.e2e.ts "custom-server-single-port.spec.ts"
    
    
     RUN  v0.25.7 /Users/naru/github.com/vitejs/vite
    
    ·
    
     Test Files  1 passed (1)
          Tests  1 passed (1)
       Start at  21:29:56
       Duration  1.21s (transform 283ms, setup 135ms, collect 33ms, tests 418ms)
</details>